src/transport/UnauthenticatedSessionTable.h

```diff
@@ -213,16 +213,18 @@ class UnauthenticatedSessionTable
     CHIP_ERROR AllocEntry(UnauthenticatedSession::SessionRole sessionRole, NodeId ephemeralInitiatorNodeID,
                           const ReliableMessageProtocolConfig & config, UnauthenticatedSession *& entry)
     {
-        entry = mEntries.CreateObject(sessionRole, ephemeralInitiatorNodeID, config);
```

Comment:
Isn't the real issue here the failure to release these objects, not the ordering of the CreateObject call and the FindLeastRecentUsedEntry?

When using non-heap pools, the pool size if fixed anyway and there should be no difference in memory use, right?  When using a heap pool, we have a leak either way, it seems like, and the right fix is to deallocate the entries when we are done with them, instead of this weird "look for the least recently used thing that now has a refcount of 0 and hence should have been deallocated"....

Was the issue here (https://github.com/project-chip/connectedhomeip/issues/29929) in the heap or non-heap case?  Seems like it was the non-heap case, but then where is the extra memory coming from?

Reply:
Yes, the issue(#29929) was the heap case, entry that CreateObject created will near be released.

Reply:
I tested the non-heap case, there is no memory leak. But the whole memory consumption will be more huge than the heap case.

Reply:
> Yes, the issue(https://github.com/project-chip/connectedhomeip/issues/29929) was the heap case

So it was a case of compiling with `CHIP_SYSTEM_CONFIG_POOL_USE_HEAP` set to 1?

If that's the case, it all makes sense, but the right answer is still that we need to make the sessions actually get released to the pool, not just leak them... This way we might limit the size of the leak to "max number of sessions we were trying to establish at a time", which is better than an unbounded leak, but still seems bad.

Reply:
Please forgive me for my previous statements not being clear enough.
The CHIP_SYSTEM_CONFIG_POOL_USE_HEAP is set to 4 in the case of issue(#29929). 

My understanding is that CHIP_SYSTEM_CONFIG_POOL_USE_HEAP only affects how many clients can use the unsecure session simultaneously, and does not affect the maximum number of CreateObject. In fact, in the heap case, there is no upper limit to CreateObject, so executing a CASE operation will create an entry, but when CASE is destroyed, the corresponding entry is not released.

Looking at the code of AllocEntry, the combination of FindLeastRecentUsedEntry and mEntries.ResetObject indicates that the designer has considered reusing the entry with the reference count of 0. However, according to the original implementation of AllocEntry, this case has no chance to be executed.

My repair solution makes it possible to reuse entries, and it doesn't have any side effects. It allows for the accumulation of entries with a reference count of 0.

Reply:
> My understanding is that CHIP_SYSTEM_CONFIG_POOL_USE_HEAP only affects how many clients can use the unsecure session simultaneously, and does not affect the maximum number of CreateObject. 

That understanding is incorrect. Heap pools do not cap the number of entries you can allocate in the pool; they just ignore that template parameter.

>  In fact, in the heap case, there is no upper limit to CreateObject, so executing a CASE operation will create an entry, but when CASE is destroyed, the corresponding entry is not released.

Yes, exactly.

> the combination of FindLeastRecentUsedEntry and mEntries.ResetObject indicates that the designer has considered reusing the entry with the reference count of 0. 

Right, because that code was written before heap pools were a thing.

> However, according to the original implementation of AllocEntry, this case has no chance to be executed.

In the heap pool case, correct.

> My repair solution makes it possible to reuse entries, and it doesn't have any side effects

What it does is help constrain the size of the leak to "max number of CASE handshakes you were ever doing in parallel", but does not fix the leak...

I created https://github.com/project-chip/connectedhomeip/pull/30025 which should fix the leak.  Would you be willing to test that @yangkqTy and let me know whether it solves things in your situation?

Reply:
OK, I will test #30025.

Reply:
I have tested #30025, it solved the issue #29929. Thanks very much for your quick response and your fix.

Reply:
@yangkqTy Thank you for finding the issue and not giving up when I asked questions about the changes!